app/controllers/measurements_controller.rb

```diff
@@ -2,47 +2,50 @@ class MeasurementsController < ApplicationController
   layout false
 
   def show
-    iteration_time = reference_time - params[:iteration].to_i * interval
-    cache_key = {
-      id: params[:id],
-      day: iteration_time.yday,
-      interval_number: interval_number(iteration_time)
-    }
-
-    @readings = Rails.cache.fetch(cache_key, expires_in: 25.hours) do
-      @measurement = Measurement.find(params[:id])
-      @measurement.
-        readings.
-        includes(:sensor).
-        where('time > ? AND time <= ?', from, to).
-        to_json(include: :sensor)
+    if params[:iteration].blank?
+      last_reading = Reading.where(measurement_id: params[:id]).last
+      readings = readings_json((last_reading&.time || Time.now),
+                               (2 * interval).minutes)
+    else
+      to = reference_time - (params[:iteration].to_i * interval).minutes
+      readings = readings_json(to, 25.hours)
     end
 
-    render json: @readings
+    render json: readings
   end
 
   private
 
+  def readings_json(to, expires_in)
+    cache_key = { id: params[:id], to: to&.to_f }
```

Comment:
I'd assume this will depend on a float-to-float comparison, on possibly very high precision, somewhere deeper. I would never trust that kind of comparison, unless I do a rounding first. Of course, the cost of a miss is not very high, it's just producing another cache entry.